### ISSUE TITLE
FFI segfault when a varargs function is given a callback as a parameter

### ISSUE THREAD

**Issue by LikeLakers2** [ffi/ffi#732]
Hi there! I'm attempting to make a FFI binding for SQLite3, and I ran into a big problem that seems to stem from FFI itself.

Specifically, I seem to be getting a segfault whenever a attached function, with `:varargs` as one of the arguments, receives a callback as one of the parameters.

The code below is the minimal code needed to replicate this. This code, when run as is, currently produces a segfault.
```ruby
require 'ffi'

module SQLite3
	extend FFI::Library
	ffi_lib 'sqlite3'
	
	callback :sqlite3_config_log_xFunc, [:pointer, :int, :string], :void
	attach_function :sqlite3_config, [:int, :varargs], :int
end

def sqlog(data, op, str)
	print 'SQLITE: '
	p [data, op, str]
end

p SQLite3.sqlite3_config(
	16, # SQLITE_CONFIG_LOG
	:sqlite3_config_log_xFunc, method(:sqlog)
)
```

To ensure this wasn't just me being stupid (or SQLite being stupid), I tried changing the parameter list for `sqlite3_config` around a bit.

When the parameter list is `[:int, :sqlite3_config_log_xFunc, :varargs]` (and the function call is edited appropriately), FFI will still segfault -- and in fact, will **always** segfault even when you don't specify the callback parameter in the function call.

However, when the parameter list is `[:int, :sqlite3_config_log_xFunc]` (with no `:varargs`), the function will work as intended. Even if I enter in invalid parameters to `sqlite3_config` (assuming their type matches up with what FFI expects the parameters to be), I no longer get a segfault from FFI, and SQLite receives the function call as expected. Except, now I've lost access to all the other config options that I'm meant to change via `sqlite3_config`, as the parameter list isn't compatible with any of the other configuration verbs offered for `sqlite3_config`.

P.S.  I am running `ffi (1.11.3 x64-mingw32)` on Windows 10, as well as SQLite 3.30.1, in case the version numbers make any difference (they shouldn't, though).

**Comment by larskanis** [comment]
There has been some trouble with callbacks on Windows in the past. Could you try if it works with older ffi versions? And with
```
ridk exec pacman -S mingw-w64-x86_64-libffi
gem inst ffi --platform=ruby -- --enable-system-libffi
```

**Comment by LikeLakers2** [comment]
Hey, @larskanis! Just wanted to let you know so you don't think I'm ignoring you, I haven't been able to test those out yet. I haven't been feeling well for some time, and wanted to wait until I could get back to feeling better before trying those out. Hope you understand!

**Comment by LikeLakers2** [comment]
@larskanis I tried installing with `--enable-system-libffi` and received the same segfault problem.

I also started going back minor versions as you suggested (installing them without `--enable-system-libffi`, and could only install as far back as FFI 1.10.0 (when I tried to install 1.9.25, it told me that my ruby version is too new). But no matter what version I used, I still got a segfault.

Sorry for the slow response, by the way!

**Comment by LikeLakers2** [comment]
Any info on this? i.e. Is it being investigated? Is there anything else I could try?

**Comment by vincentisambart** [comment]
Looking at the source code, it seems FFI currently indeed does not support passing callbacks via varargs.
A variadic invocation always passes `NULL` and `0` for `callbackParameters` and `callbackCount` when calling `rbffi_SetupCallParams`.
https://github.com/ffi/ffi/blob/3f862ca78ec4e6c2a5eabca8a5f7780d1774c1a0/ext/ffi_c/Variadic.c#L250-L251

If the arguments include a callback,`rbffi_SetupCallParams` does not expect `callbackParameters` to be `NULL` and crashes when trying to dereference `callbackParameters` at the following line.
https://github.com/ffi/ffi/blob/3f862ca78ec4e6c2a5eabca8a5f7780d1774c1a0/ext/ffi_c/Call.c#L317

It seems that `variadic_invoke` should at least pass correct `callbackParameters` and `callbackCount` to `rbffi_SetupCallParams` (not sure if that would be enough though).

**Comment by vincentisambart** [comment]
This should be fixed in ffi 1.15.0.
@LikeLakers2 Could you check?

**Comment by LikeLakers2** [comment]
@vincentisambart Unfortunately, I don't use Ruby anymore, nor do I have it installed; but the code in the original post should be enough to see if it's fixed. If you need a SQLite binary to use with that code, SQLite offers precompiled libraries: https://sqlite.org/download.html

**Comment by larskanis** [comment]
Callback as `:varargs` argument is implemented in #885 . So this should be fixed in ffi-1.15.0.
